[Bug]: aws_ecs_express_gateway_service - Provider produced inconsistent result after apply due to network_configuration.security_groups being   modified by Express Mode

### Terraform and AWS Provider Version

```console
Terraform v1.10
  provider registry.terraform.io/hashicorp/aws v6.38.0 (or higher)
```

### Affected Resource(s) or Data Source(s)

aws_ecs_express_gateway_service

### Expected Behavior

The provider should successfully save state after creating the service. Express Mode
  automatically adds its own ALB security group to network_configuration.security_groups                                      
  during creation — the provider should account for this and not fail when the returned
  state differs from the planned value.

### Actual Behavior

The resource is created successfully in AWS but the provider fails with
  "Provider produced inconsistent result after apply" because Express Mode adds its own                                       
  security group to network_configuration.security_groups during creation, changing the
  value from null (or 1 custom SG) to a set containing Express Mode's managed SG.
  Terraform cannot save the state, leaving the resource untracked and causing
  "Resource Already Exists" errors on subsequent applies.

### Relevant Error/Panic Output

Error: Provider produced inconsistent result after apply

  When applying changes to
  module.lactools_ecs_express[0].aws_ecs_express_gateway_service.main,
  provider "provider[\"registry.terraform.io/hashicorp/aws\"]" produced an
  unexpected new value: .network_configuration[0].security_groups: was null,
  but now cty.SetVal([]cty.Value{cty.StringVal("sg-00185ff5c2afc3f61")}).

  This is a bug in the provider, which should be reported in the provider's
  own issue tracker.


### Sample Terraform Configuration

<details open>
<summary>Click to expand configuration</summary>

```hcl
resource "aws_ecs_express_gateway_service" "main" {
    service_name = "my-service"
    cpu          = "1024"
    memory       = "2048"
  
    execution_role_arn      = aws_iam_role.execution.arn
    infrastructure_role_arn = aws_iam_role.infrastructure.arn
    task_role_arn           = aws_iam_role.task.arn

    health_check_path = "/health"

    primary_container {
      image          = "123456789012.dkr.ecr.us-east-2.amazonaws.com/myapp:latest"
      container_port = 80

      aws_logs_configuration {  
        log_group         = "/ecs/my-service"
        log_stream_prefix = "my-service"
      }
    }
  
    # Reproduced with security_groups specified:
    network_configuration {
      subnets         = ["subnet-aaa", "subnet-bbb"]
      security_groups = ["sg-xxx"]
    }
  
    # Also reproduced without security_groups (null):
    # network_configuration {
    #   subnets = ["subnet-aaa", "subnet-bbb"]
    # }
  
    scaling_target {
      auto_scaling_metric       = "AVERAGE_CPU"
      auto_scaling_target_value = 70
      min_task_count            = 1
      max_task_count            = 4
    }
  }
```
</details>


### Steps to Reproduce

1. Create an aws_ecs_express_gateway_service resource with any network_configuration
     (with or without security_groups).                                                                                       

  2. Run terraform apply.

  3. Wait for the service to be created in AWS (~9 minutes).
  
  4. Observe the "Provider produced inconsistent result after apply" error.
  
  5. Verify in the AWS console or CLI that the service WAS successfully created
     (status ACTIVE or INACTIVE) but is not tracked in Terraform state.

  6. Run terraform apply again — it will fail with "Resource Already Exists" because
     the service exists in AWS but not in state.


### Debug Logging

Not available — the error occurs in a GitLab CI pipeline. The relevant error output
  is already captured in the "Relevant Error/Panic Output" section above.

### GenAI / LLM Assisted Development

Kiro

### Important Facts and References

- The service is created in a standard AWS commercial region (us-east-2).
  - Express Mode automatically adds its own ALB security group to
    network_configuration.security_groups during creation. This is by design
    per AWS documentation, but the Terraform provider does not account for it.
  - Adding `lifecycle { ignore_changes = [network_configuration[0].security_groups] }`
    does NOT fix the issue — the inconsistency check happens during apply before
    state is saved, not during planning.
 

### Would you like to implement a fix?

No

I reproduced this issue by running the existing acceptance test without any modification:
```console
$ make testacc TESTS='TestAccECSExpressGatewayService_networkConfiguration' PKG=ecs       
make: Verifying source code with gofmt...
==> Checking that code complies with gofmt requirements...
make: Validating schemas
ok      github.com/hashicorp/terraform-provider-aws/internal/provider/sdkv2     5.617s
ok      github.com/hashicorp/terraform-provider-aws/internal/provider/framework 5.622s
make: Running acceptance tests on branch: 🌿 f-aws_acm_certificate-add_created_at 🌿...
TF_ACC=1 go1.26.3 test ./internal/service/ecs/... -v -count 1 -parallel 20 -run='TestAccECSExpressGatewayService_networkConfiguration'  -timeout 360m -vet=off -buildvcs=false
2026/05/15 07:10:25 Creating Terraform AWS Provider (SDKv2-style)...
2026/05/15 07:10:25 Initializing Terraform AWS Provider (SDKv2-style)...
=== RUN   TestAccECSExpressGatewayService_networkConfiguration
=== PAUSE TestAccECSExpressGatewayService_networkConfiguration
=== CONT  TestAccECSExpressGatewayService_networkConfiguration
    express_gateway_service_test.go:360: Step 1/1 error: After applying this test step, the refresh plan was not empty.
        stdout
        
        
        Terraform used the selected providers to generate the following execution
        plan. Resource actions are indicated with the following symbols:
          ~ update in-place
        
        Terraform will perform the following actions:
        
          # aws_ecs_express_gateway_service.test will be updated in-place
          ~ resource "aws_ecs_express_gateway_service" "test" {
              ~ current_deployment      = "arn:aws:ecs:us-west-2:123456789012:service-deployment/default/tf-acc-test-7703478190532023843-service/Bj5S9O-e3Qvsbup_Rh1BY" -> (known after apply)
              ~ ingress_paths           = [
                  - {
                      - access_type = "PRIVATE"
                      - endpoint    = "tf-3a43098a9a47407ab07f39235ccc2d0a.ecs.us-west-2.on.aws"
                    },
                ] -> (known after apply)
              ~ network_configuration   = [
                  ~ {
                      ~ security_groups = [
                          - "sg-00295ddde226cb7b3",
                            # (1 unchanged element hidden)
                        ]
                        # (1 unchanged attribute hidden)
                    },
                ]
              ~ service_revision_arn    = "arn:aws:ecs:us-west-2:123456789012:service-revision/default/tf-acc-test-7703478190532023843-service/9309441587984617189" -> (known after apply)
                # (13 unchanged attributes hidden)
        
                # (1 unchanged block hidden)
            }
        
        Plan: 0 to add, 1 to change, 0 to destroy.
--- FAIL: TestAccECSExpressGatewayService_networkConfiguration (425.73s)

```
I will try to fix this issue.

--- a/.changelog/47944.txt
+++ b/.changelog/47944.txt
@@ -1,3 +1,3 @@
 ```release-note:bug
-resource/aws_ecs_express_gateway_service: Fix "inconsistent result after apply" error for `network_configuration[0].security_groups` when using `network_configuration`
+resource/aws_ecs_express_gateway_service: Fix "inconsistent result after apply" error for `network_configuration[0].security_groups` when using `network_configuration`. `ec2:DescribeSecurityGroups` IAM permission is newly required.
 ```

--- a/.changelog/47944.txt
+++ b/.changelog/47944.txt
@@ -1,0 +1,3 @@
+```release-note:bug
+resource/aws_ecs_express_gateway_service: Fix "inconsistent result after apply" error for `network_configuration[0].security_groups` when using `network_configuration`
+```

--- a/internal/service/ecs/express_gateway_service.go
+++ b/internal/service/ecs/express_gateway_service.go
@@ -16,6 +16,8 @@ import (
 	"github.com/YakDriver/smarterr"
 	"github.com/aws/aws-sdk-go-v2/aws"
 	"github.com/aws/aws-sdk-go-v2/aws/arn"
+	"github.com/aws/aws-sdk-go-v2/service/ec2"
+	ec2types "github.com/aws/aws-sdk-go-v2/service/ec2/types"
 	"github.com/aws/aws-sdk-go-v2/service/ecs"
 	awstypes "github.com/aws/aws-sdk-go-v2/service/ecs/types"
 	"github.com/hashicorp/terraform-plugin-framework-timeouts/resource/timeouts"
@@ -37,6 +39,7 @@ import (
 	fwflex "github.com/hashicorp/terraform-provider-aws/internal/framework/flex"
 	fwtypes "github.com/hashicorp/terraform-provider-aws/internal/framework/types"
 	"github.com/hashicorp/terraform-provider-aws/internal/retry"
+	tfec2 "github.com/hashicorp/terraform-provider-aws/internal/service/ec2"
 	"github.com/hashicorp/terraform-provider-aws/internal/smerr"
 	tftags "github.com/hashicorp/terraform-provider-aws/internal/tags"
 	"github.com/hashicorp/terraform-provider-aws/internal/tfresource"
@@ -278,6 +281,7 @@ func (r *expressGatewayServiceResource) Create(ctx context.Context, req resource
 	if len(waitOut.ActiveConfigurations) > 0 {
 		// Save plan's env/secret ordering before flattening (API may reorder).
 		planEnv, planSecrets := preservePlanContainerOrdering(ctx, plan.PrimaryContainer)
+		planNetworkConfiguration := plan.NetworkConfiguration
 
 		smerr.AddEnrich(ctx, &resp.Diagnostics, fwflex.Flatten(ctx, waitOut.ActiveConfigurations[0], &plan))
 		if resp.Diagnostics.HasError() {
@@ -285,6 +289,7 @@ func (r *expressGatewayServiceResource) Create(ctx context.Context, req resource
 		}
 
 		restorePlanContainerOrdering(ctx, &plan.PrimaryContainer, planEnv, planSecrets)
+		restorePlanNetworkConfigurationSecurityGroups(ctx, &plan.NetworkConfiguration, planNetworkConfiguration)
 	}
 
 	plan.Cluster = fwflex.StringValueToFramework(ctx, cluster)
@@ -303,6 +308,7 @@ func (r *expressGatewayServiceResource) Read(ctx context.Context, req resource.R
 	}
 
 	conn := r.Meta().ECSClient(ctx)
+	ec2Conn := r.Meta().EC2Client(ctx)
 
 	serviceARN := fwflex.StringValueFromFramework(ctx, state.ServiceARN)
 	out, err := findExpressGatewayServiceByARN(ctx, conn, serviceARN)
@@ -333,6 +339,8 @@ func (r *expressGatewayServiceResource) Read(ctx context.Context, req resource.R
 		if resp.Diagnostics.HasError() {
 			return
 		}
+
+		filterAmazonECSManagedSecurityGroups(ctx, ec2Conn, &state.NetworkConfiguration)
 
 		// Restore state ordering if env vars are unchanged (no-op during import).
 		restoreContainerOrderingIfUnchanged(ctx, &state.PrimaryContainer, stateEnv, stateSecrets)
@@ -440,6 +448,7 @@ func (r *expressGatewayServiceResource) Update(ctx context.Context, req resource
 	if len(waitOut.ActiveConfigurations) > 0 {
 		// Save plan's env/secret ordering before flattening (API may reorder).
 		planEnv, planSecrets := preservePlanContainerOrdering(ctx, plan.PrimaryContainer)
+		planNetworkConfiguration := plan.NetworkConfiguration
 
 		orderExpressGatewayContainerEnvironmentVariables(&waitOut.ActiveConfigurations[0])
 		smerr.AddEnrich(ctx, &resp.Diagnostics, fwflex.Flatten(ctx, waitOut.ActiveConfigurations[0], &plan))
@@ -448,6 +457,7 @@ func (r *expressGatewayServiceResource) Update(ctx context.Context, req resource
 		}
 
 		restorePlanContainerOrdering(ctx, &plan.PrimaryContainer, planEnv, planSecrets)
+		restorePlanNetworkConfigurationSecurityGroups(ctx, &plan.NetworkConfiguration, planNetworkConfiguration)
 	}
 
 	// Set Optional+Computed attributes from API response
@@ -859,6 +869,81 @@ func restoreContainerOrderingIfUnchanged(
 		return
 	}
 	*primaryContainer = updated
+}
+
+func restorePlanNetworkConfigurationSecurityGroups(
+	ctx context.Context,
+	networkConfiguration *fwtypes.ListNestedObjectValueOf[expressGatewayServiceNetworkConfigurationModel],
+	planNetworkConfiguration fwtypes.ListNestedObjectValueOf[expressGatewayServiceNetworkConfigurationModel],
+) {
+	planConfig, diags := planNetworkConfiguration.ToPtr(ctx)
+	if diags.HasError() || planConfig == nil || planConfig.SecurityGroups.IsUnknown() {
+		return
+	}
+
+	currentConfig, diags := networkConfiguration.ToPtr(ctx)
+	if diags.HasError() || currentConfig == nil {
+		return
+	}
+
+	currentConfig.SecurityGroups = planConfig.SecurityGroups
+
+	updated, diags := fwtypes.NewListNestedObjectValueOfPtr(ctx, currentConfig)
+	if diags.HasError() {
+		return
+	}
+
+	*networkConfiguration = updated
+}
+
+func filterAmazonECSManagedSecurityGroups(
+	ctx context.Context,
+	conn *ec2.Client,
+	networkConfiguration *fwtypes.ListNestedObjectValueOf[expressGatewayServiceNetworkConfigurationModel],
+) {
+	currentConfig, diags := networkConfiguration.ToPtr(ctx)
+	if diags.HasError() || currentConfig == nil || currentConfig.SecurityGroups.IsNull() || currentConfig.SecurityGroups.IsUnknown() {
+		return
+	}
+
+	securityGroups := fwflex.ExpandFrameworkStringValueSet(ctx, currentConfig.SecurityGroups)
+	if len(securityGroups) == 0 {
+		return
+	}
+
+	groups, err := tfec2.FindSecurityGroups(ctx, conn, &ec2.DescribeSecurityGroupsInput{
+		GroupIds: securityGroups,
+	})
+	if err != nil {
+		return
+	}
+
+	filteredGroupIDs := make([]string, 0, len(groups))
+	for _, group := range groups {
+		if !hasAmazonECSManagedTag(group.Tags) {
+			filteredGroupIDs = append(filteredGroupIDs, aws.ToString(group.GroupId))
+		}
+	}
+
+	slices.Sort(filteredGroupIDs)
+	currentConfig.SecurityGroups = fwflex.FlattenFrameworkStringValueSetOfString(ctx, filteredGroupIDs)
+
+	updated, diags := fwtypes.NewListNestedObjectValueOfPtr(ctx, currentConfig)
+	if diags.HasError() {
+		return
+	}
+
+	*networkConfiguration = updated
+}
+
+func hasAmazonECSManagedTag(tags []ec2types.Tag) bool {
+	for _, tag := range tags {
+		if aws.ToString(tag.Key) == "AmazonECSManaged" && aws.ToString(tag.Value) == "true" {
+			return true
+		}
+	}
+
+	return false
 }
 
 // envListsEquivalent returns true if two environment variable lists contain the same

--- a/internal/service/ecs/express_gateway_service.go
+++ b/internal/service/ecs/express_gateway_service.go
@@ -11,6 +11,7 @@ import (
 	"errors"
 	"fmt"
 	"slices"
+	"strings"
 	"time"
 
 	"github.com/YakDriver/smarterr"
@@ -291,6 +292,7 @@ func (r *expressGatewayServiceResource) Create(ctx context.Context, req resource
 		restorePlanContainerOrdering(ctx, &plan.PrimaryContainer, planEnv, planSecrets)
 		restorePlanNetworkConfigurationSecurityGroups(ctx, &plan.NetworkConfiguration, planNetworkConfiguration)
 	}
+	normalizeIngressPathEndpoints(ctx, &plan.IngressPaths)
 
 	plan.Cluster = fwflex.StringValueToFramework(ctx, cluster)
 	plan.CurrentDeployment = fwflex.StringToFramework(ctx, waitOut.CurrentDeployment)
@@ -345,6 +347,7 @@ func (r *expressGatewayServiceResource) Read(ctx context.Context, req resource.R
 		// Restore state ordering if env vars are unchanged (no-op during import).
 		restoreContainerOrderingIfUnchanged(ctx, &state.PrimaryContainer, stateEnv, stateSecrets)
 	}
+	normalizeIngressPathEndpoints(ctx, &state.IngressPaths)
 
 	// Set Optional+Computed attributes from API response
 	if out.Cluster != nil {
@@ -459,6 +462,7 @@ func (r *expressGatewayServiceResource) Update(ctx context.Context, req resource
 		restorePlanContainerOrdering(ctx, &plan.PrimaryContainer, planEnv, planSecrets)
 		restorePlanNetworkConfigurationSecurityGroups(ctx, &plan.NetworkConfiguration, planNetworkConfiguration)
 	}
+	normalizeIngressPathEndpoints(ctx, &plan.IngressPaths)
 
 	// Set Optional+Computed attributes from API response
 	if waitOut.Cluster != nil {
@@ -944,6 +948,45 @@ func hasAmazonECSManagedTag(tags []ec2types.Tag) bool {
 	}
 
 	return false
+}
+
+func normalizeIngressPathEndpoints(ctx context.Context, ingressPaths *fwtypes.ListNestedObjectValueOf[ingressPathSummaryModel]) {
+	paths, diags := ingressPaths.ToSlice(ctx)
+	if diags.HasError() || len(paths) == 0 {
+		return
+	}
+
+	updatedPaths := make([]*ingressPathSummaryModel, 0, len(paths))
+	changed := false
+
+	for _, ingressPath := range paths {
+		if ingressPath == nil || ingressPath.Endpoint.IsNull() || ingressPath.Endpoint.IsUnknown() {
+			updatedPaths = append(updatedPaths, ingressPath)
+			continue
+		}
+
+		endpoint := ingressPath.Endpoint.ValueString()
+		if strings.HasPrefix(endpoint, "https://") || endpoint == "" {
+			updatedPaths = append(updatedPaths, ingressPath)
+			continue
+		}
+
+		clone := *ingressPath
+		clone.Endpoint = types.StringValue("https://" + endpoint)
+		updatedPaths = append(updatedPaths, &clone)
+		changed = true
+	}
+
+	if !changed {
+		return
+	}
+
+	updated, diags := fwtypes.NewListNestedObjectValueOfSlice(ctx, updatedPaths, nil)
+	if diags.HasError() {
+		return
+	}
+
+	*ingressPaths = updated
 }
 
 // envListsEquivalent returns true if two environment variable lists contain the same

--- a/internal/service/ecs/express_gateway_service.go
+++ b/internal/service/ecs/express_gateway_service.go
@@ -342,7 +342,7 @@ func (r *expressGatewayServiceResource) Read(ctx context.Context, req resource.R
 			return
 		}
 
-		filterAmazonECSManagedSecurityGroups(ctx, ec2Conn, &state.NetworkConfiguration)
+		filterManagedSecurityGroups(ctx, ec2Conn, &state.NetworkConfiguration)
 
 		// Restore state ordering if env vars are unchanged (no-op during import).
 		restoreContainerOrderingIfUnchanged(ctx, &state.PrimaryContainer, stateEnv, stateSecrets)
@@ -900,7 +900,7 @@ func restorePlanNetworkConfigurationSecurityGroups(
 	*networkConfiguration = updated
 }
 
-func filterAmazonECSManagedSecurityGroups(
+func filterManagedSecurityGroups(
 	ctx context.Context,
 	conn *ec2.Client,
 	networkConfiguration *fwtypes.ListNestedObjectValueOf[expressGatewayServiceNetworkConfigurationModel],
@@ -924,7 +924,7 @@ func filterAmazonECSManagedSecurityGroups(
 
 	filteredGroupIDs := make([]string, 0, len(groups))
 	for _, group := range groups {
-		if !hasAmazonECSManagedTag(group.Tags) {
+		if !hasManagedTag(group.Tags) {
 			filteredGroupIDs = append(filteredGroupIDs, aws.ToString(group.GroupId))
 		}
 	}
@@ -940,7 +940,7 @@ func filterAmazonECSManagedSecurityGroups(
 	*networkConfiguration = updated
 }
 
-func hasAmazonECSManagedTag(tags []ec2types.Tag) bool {
+func hasManagedTag(tags []ec2types.Tag) bool {
 	for _, tag := range tags {
 		if aws.ToString(tag.Key) == "AmazonECSManaged" && aws.ToString(tag.Value) == "true" {
 			return true

--- a/internal/service/ecs/express_gateway_service.go
+++ b/internal/service/ecs/express_gateway_service.go
@@ -971,9 +971,8 @@ func normalizeIngressPathEndpoints(ctx context.Context, ingressPaths *fwtypes.Li
 			continue
 		}
 
-		clone := *ingressPath
-		clone.Endpoint = types.StringValue("https://" + endpoint)
-		updatedPaths = append(updatedPaths, &clone)
+		ingressPath.Endpoint = types.StringValue("https://" + endpoint)
+		updatedPaths = append(updatedPaths, ingressPath)
 		changed = true
 	}
 

